### ISSUE TITLE
Wrong silhouette generated of a cylinder

### ISSUE THREAD

**Issue by asmwarrior** [vectorgraphics/asymptote#117]
Hi, I reported this issue in the Sourceforge forum, but as suggested by John, it should report the bug here, see [Wrong silhouette generated of a cylinder](https://sourceforge.net/p/asymptote/discussion/409349/thread/930e874b45/).

The test code is:

~~~
import settings;
settings.interactiveView=true;
settings.render=0;
import solids;

size3(200,IgnoreAspect);

currentprojection=perspective(camera=(0,0,-10),up=(0,1,0),target=(0,0,0),autoadjust=false,center=false);

revolution r=cylinder((0,3,4),1.5,2,Y);

draw(r,green+2pt);
~~~

And the result image is not correct. Some arcs should be solid because they belongs to the "front" of the skeleton.

**Comment by asmwarrior** [comment]
Hi, I take some time to debug this issue. I think I have some good catches.

In the code snippet:

https://github.com/vectorgraphics/asymptote/blob/1812df7829f5f7fcda6cdca25acedd9b97b3145d/base/solids.asy#L127-L140

Especially the line:
https://github.com/vectorgraphics/asymptote/blob/1812df7829f5f7fcda6cdca25acedd9b97b3145d/base/solids.asy#L134

The first dot product:
`dot(axis,camera-P.target)` looks like that we try to see whether the base(either the top or the bottom base) of the cylinder will be seen from the camera.

But I'm not sure why we use `camera-P.target`? I think the better choice is `dot(axis,camera-relpoint(g,t))`.

With this changes, I see my example test code in https://github.com/vectorgraphics/asymptote/issues/117#issue-503029354 shown correctly.

I also try to move the cylinder up and down along the Y axis(such as I can change the line to `revolution r=cylinder((0,0,4),1.5,2,Y);` or `revolution r=cylinder((0,-3,4),1.5,2,Y);`, and it all shows correctly.

Any ideas?
I'm not good at asymptote programming language debugging, all what I can do is adding many `write` functions, like:

```
    write("axis=", axis);
    write("projection.target=",P.target);
    write("projection.camera=",P.camera);
    write("camera-P.target=", camera-P.target);
    write("dot(axis,camera-P.target)=",dot(axis,camera-P.target));
    write("dot(axis,dir(g,midtime))=",dot(axis,dir(g,midtime)));
    write("sign=",sign);
    write("t=",t);
    write("M=",M);
    write("m=",m);
    write("epsilon=",epsilon);
```

Thanks.

**Comment by johncbowman** [comment]
Thanks for the suggestion. The problem indeed has to do with this test, but that change breaks other examples, like the one below. This code needs a careful rewrite.

```
import solids;
unitsize(cm);
currentprojection = orthographic(5,4,2);
revolution r=cylinder(O,1,1,Z);
draw(r,blue+thick());
draw(surface(r),red);
```

**Comment by asmwarrior** [comment]
Hi, John, I think there is no issue here.
First, this code works OK:

```
import settings;
settings.interactiveView=true;
settings.render=0;
import solids;
unitsize(cm);
currentprojection = orthographic(5,4,2);
revolution r=cylinder(O,1,1,Z);
draw(surface(r),red);
draw(r,blue+thick());
```
Here is the screen shot:
[render0-correct.pdf](https://github.com/vectorgraphics/asymptote/files/3783890/render0-correct.pdf)
In the above example, I first draw the surface, and then the wire-frame.

Now, if I exchange the drawing sequence
```
import settings;
settings.interactiveView=true;
settings.render=0;
import solids;
unitsize(cm);
currentprojection = orthographic(5,4,2);
revolution r=cylinder(O,1,1,Z);
draw(r,blue+thick());
draw(surface(r),red);
```
I get the top base solid line get covered by the surface, see the image here:
[render0-bad.pdf](https://github.com/vectorgraphics/asymptote/files/3783911/render0-bad.pdf)
But I think this is the expected behavior, since we use the render=0.

Finally, If I use the OpenGL based render, the result image shows correctly either the wire-frame or the surface draw first. I think this is because the wire-frame is actually a pipe which has a fixed width, so the surface never cover the cover the whole wire.

**Comment by asmwarrior** [comment]
Oh, you are correct, I see the issue in your link(WebGL). In-fact, the issue also happens in my previous attached pdf files. Sorry about that. Hope I can take some time to debug this issue again.

About the settings.render=0. I think it still has its usage . I sometimes need those vector 3D images in the PDF. Not every people use the the acrobat pdf reader which supports PRC.

**Comment by asmwarrior** [comment]
Hi, John, I found the reason why your sample code shows wrong top lines.

The problem is the camera position is not the "5,4,2" when calculating the dot product.
In the function `void transverse(skeleton s, real t, int n=nslice, projection P) `. I just print the variable `P.camera`, and the `camera`(which is from the function call `triple camera=camera(P);`, they have different values:

```
projection.camera=(5,4,2)
camera=(1.78885438199983,1.43108350559987,0.715541752799933)
```
I see there is a function:

```
  triple camera(projection P) {
    triple camera=P.camera;
    if(P.infinity) {
      real s=abs(M-m)+abs(m-P.target);
      camera=P.target+camerafactor*s*unit(P.vector());
    }
    return camera;
  }
```
Now, you see the dot product: `dot(axis,camera-P.target)` or my suggested dot product:
`dot(axis,camera-relpoint(g,t))`. Here, the **camera** position is a bit lower of the top cap of the cylinder in the Z axis, so we get the wrong dot result.

I think this only affects the orthographic projection, since for perspective projection, the `camera == P.camera`.

Any ideas?

BTW: I'm not sure what does the variable M and m means.

**Comment by johncbowman** [comment]
Yes, I noticed this too. The purpose of the camera function is to move the orthographic camera (for which only the direction matters) further away, outside of the bounding cube with vertices m and M.
But obviously this isn't working as it should.

**Comment by johncbowman** [comment]
Fixed in 4dc96555e0f56e387d4f64d26865d1cfdcefc0fd. I will hold off on the 2.60 release for one more day so that this can be tested by others. To try it out, grab a new copy of
https://raw.githubusercontent.com/vectorgraphics/asymptote/HEAD/base/solids.asy
